File: appletv/Sources/HexStackerKit/Game/PadSeats.swift
import Foundation

/// Gamepads attached to the TV itself, as players.
///
/// A pad is not a second kind of player. Every press becomes the SAME message a
/// phone would have sent and goes through the coordinator's inbound path, so
/// joining, auto-naming, colour slots, host election, liveness and pause all keep
/// running their one implementation. What a local seat skips is the relay, which
/// is why its peer index comes from a range the relay will never hand out — see
/// `localSeatBase` below for the range and the packed-frame constraint that makes
/// it a HIGH POSITIVE one. `isLocalSeat` is what guards the `transport.sendTo`
/// call sites.
///
/// The MAPPING is not here. Which button rotates, how fast a held direction
/// repeats and how the stick scales a soft drop all live in `server/PadMapper.js`
/// and cross through `EngineBridge.padPoll`, because the web display, this and
/// Android TV must agree about what a press means. What IS here is the part that
/// needs a shell: which pads exist, which seat each one holds, and where a menu
/// press goes.
///
/// This type deliberately does not import GameController. Reading the hardware is
/// a `PadSource`, injected, so the seat lifecycle and the menu routing below are
/// testable without a controller plugged into the machine.
public protocol PadSource {
    /// Attached pads in STABLE slot order: a pad must keep its slot across polls,
    /// and reclaim it after a replug, or it would be a new player every time.
    func pads() -> [PadReading]
    /// Effects a phone gets as haptics through its own vibrate() call.
    func rumble(slot: Int, durationMs: Double, weak: Double, strong: Double)
}

public struct PadReading {
    public let slot: Int
    /// The product string, cleaned up through the shared rules into a name.
    public let id: String
    /// W3C "standard" mapping order, so one mapper serves every platform.
    public let buttons: [Bool]
    /// `[leftX, leftY]`, y POSITIVE DOWN as the web reports it. The mapper reads
    /// nothing past the left stick, so nothing else is carried.
    public let axes: [Double]

    public init(slot: Int, id: String, buttons: [Bool], axes: [Double]) {
        self.slot = slot
        self.id = id
        self.buttons = buttons
        self.axes = axes
    }
}

/// Button indices this file routes by hand. The rest are the mapper's business
/// and never named here. Kept in sync with `PAD_BTN` in server/PadMapper.js.
enum PadButton {
    static let faceDown = 0
    static let l1 = 4
    static let r1 = 5
    static let l2 = 6
    static let r2 = 7
    static let start = 9
}

public final class PadSeats {
    /// An id the relay will never hand out (it owns slot 0 and gives out 1...MAX).
    /// Mirrors LOCAL_SEAT_BASE in server/PadMapper.js, which owns the definition
    /// and the reason it must be POSITIVE: a player id crosses to us inside
    /// PartyCore's packed frame, where every integer is one UTF-16 code unit, and
    /// a negative one is unencodable — packFrame throws, so the first frame of a
    /// match with a pad seated kills the game. Pinned by
    /// tests/protocol-swift-parity.test.js.
    static let localSeatBase = 900

    /// Derived from the pad's own slot, so unplugging and replugging the same pad
    /// lands back on the same seat: a reconnect, not a new player.
    public static func seatId(forSlot slot: Int) -> Int { localSeatBase + slot }

    /// True for a seat this display owns rather than one the relay handed out.
    /// Every `sendTo` has to ask, because there is nothing on the wire to send to.
    public static func isLocalSeat(_ peerIndex: Int) -> Bool { peerIndex >= localSeatBase }

    private let source: PadSource
    private unowned let coordinator: DisplayCoordinator
    /// slot -> seat id, for pads that have actually joined.
    private var seated: [Int: Int] = [:]
    /// slot -> pad-clock time of a REFUSED join (room full). Because tvOS joins
    /// on connect rather than on a press, a refused pad would otherwise re-send
    /// HELLO on every poll — three bridge crossings per frame, forever. There is
    /// no press edge to retry on here, so the retry is paced by time instead.
    private var refusedAt: [Int: Double] = [:]
    private let refusedRetryMs: Double = 1000

    public init(source: PadSource, coordinator: DisplayCoordinator) {
        self.source = source
        self.coordinator = coordinator
    }

    /// Seats currently held by a pad, which the shell needs in order to know
    /// whether to suppress focus (see `poll`).
    public var hasSeats: Bool { !seated.isEmpty }

    /// Seats that joined on the LAST collect. Their press still goes to the
    /// mapper, which is the point: it becomes the baseline, so the button that
    /// joined reads as already-down rather than as a fresh press. What must not
    /// happen is acting on it, or the bottom face button would join and start
    /// the round in one press. Spans collect to route, which is why it is not a
    /// local.
    private var joinedNow: Set<Int> = []

    /// One poll's pad states: retire vanished pads, seat new ones, stamp
    /// liveness. The mapping itself is the shim's; the CALLER decides which
    /// crossing carries it — poll's own, or the playing tick's frame
    /// (EngineBridge.framePads), which is what keeps a playing tick at one
    /// evaluate. `nowMs` is the pad clock, NOT wall time.
    public func collectStates(nowMs: Double) -> [EngineBridge.PadState] {
        let readings = source.pads()
        retireVanished(readings)
        validateSeats(nowMs: nowMs)

        // Join before mapping: a pad that has not joined has no seat to attribute
        // input to, and the joining press must not also fire whatever it is bound
        // to (see joinedNow above).
        joinedNow.removeAll()
        var states: [EngineBridge.PadState] = []
        for reading in readings {
            let existing = seated[reading.slot] != nil
            guard let seat = seat(for: reading, nowMs: nowMs) else { continue }
            if !existing { joinedNow.insert(seat) }
            // A local seat sends nothing over the wire, so nothing else proves it
            // is still there. The pad being present in this poll IS the proof;
            // without it the liveness sweep would expire an idle player mid-game.
            coordinator.markLocalSeatSeen(seat)
            states.append(EngineBridge.PadState(
                seat: seat, buttons: reading.buttons, axes: reading.axes))
        }
        return states
    }

    /// Route one poll's results — the menu edges and the rumble flag; the game
    /// input never comes back (the shim feeds it to the engine itself).
    public func route(_ results: [EngineBridge.PadResult], playing: Bool) {
        for result in results {
            // The joining press is a baseline, never an action. See joinedNow.
            if joinedNow.contains(result.seat) { continue }
            // The one effect driven by what the PLAYER did rather than what
            // happened to them. Fired off the mapping rather than the resulting
            // lock event, so the thump lands with the press.
            if result.hardDrop { rumble(seat: result.seat, "hardDrop") }
            guard !playing else {
                if result.pressed.contains(PadButton.start) {
                    coordinator.deliverLocal(from: result.seat, data: ["type": MSG.pauseGame])
                }
                continue
            }
            // Index 9 is bound here on EVERY screen, not only while the pad owns
            // input, because tvOS never puts a gamepad's Menu button on the
            // responder chain — it stays in GameController. That is why the remote
            // exits the lobby and a pad does not: the code path is identical, so
            // the press simply is not arriving. Leaving it to `handleMenu` (as the
            // face buttons are left to the focus engine) means nothing hears it,
            // which is what stranded the pause overlay: Menu raised it and Menu
            // could not put it back.
            // Countdown counts as well as playing, the same pair `remoteTogglePause`
            // accepts and the same pair the web allows — otherwise the remote can
            // pause the 3-2-1 and a pad cannot, for no reason a player could guess.
            if result.pressed.contains(PadButton.start),
               coordinator.state == .playing || coordinator.state == .countdown {
                coordinator.remoteTogglePause()
                continue
            }
            for direction in result.nav { onMenuNav(seat: result.seat, direction: direction) }
            for index in result.pressed { onMenuPress(seat: result.seat, index: index) }
        }
    }

    /// Collect, map and route in one go — the path for every tick that is NOT
    /// running a frame (lobby, results, countdown, paused). The playing tick
    /// rides its mapping on the frame crossing instead (see collectStates).
    public func poll(nowMs: Double, playing: Bool) {
        let states = collectStates(nowMs: nowMs)
        guard !states.isEmpty else { return }

        let results: [EngineBridge.PadResult]
        do { results = try coordinator.padPoll(states, nowMs: nowMs, playing: playing) }
        catch {
            FileHandle.standardError.write(Data("[pad] poll failed: \(error)\n".utf8))
            return
        }
        route(results, playing: playing)
    }

    // MARK: - Seat lifecycle

    /// A row can disappear without its pad going anywhere: a session reset
    /// clears the whole roster, and a phone claiming the seat's rejoin QR moves
    /// the row to its own index. Give the seat up so the next press joins fresh
    /// instead of feeding a player that is gone. Swept at 1Hz, not per frame:
    /// `roster()` is a bridge crossing, and the fused playing tick exists to
    /// keep those at one per frame (a doomed press inside the window is a no-op
    /// everywhere — the engine and the room both ignore unknown players).
    /// Android reads its native roster mirror per frame instead; tvOS keeps no
    /// mirror, and a 1s window costs nothing observable.
    private var seatSweepMs: Double = -.infinity

    private func validateSeats(nowMs: Double) {
        guard !seated.isEmpty, nowMs - seatSweepMs >= 1000 else { return }
        seatSweepMs = nowMs
        let rows = Set(coordinator.roster().map(\.peerIndex))
        for (slot, seat) in seated where !rows.contains(seat) {
            seated.removeValue(forKey: slot)
        }
    }

    private func seat(for reading: PadReading, nowMs: Double) -> Int? {
        if let seat = seated[reading.slot] {
            return seat
        }
        // CONNECTING joins, with no press required — the one place tvOS differs
        // from the web and Android, and it follows from the platform rather than
        // from taste. Menus here belong to the focus engine, which means a pad's
        // first press also clicks whatever is highlighted; making that press the
        // join is what had a controller take a seat and start the round together.
        // There is no press to disarm, because there is no joining press.
        //
        // What makes it safe is that tvOS drops an idle pad from
        // `GCController.controllers()` when it sleeps, so "connected" already
        // means "awake and in someone's hands" — the reason the web has to wait
        // for a button (`navigator.getGamepads()` reports nothing until then) does
        // not apply. A pad left in a drawer is not connected, so it cannot join.
        if let at = refusedAt[reading.slot], nowMs - at < refusedRetryMs { return nil }
        return join(reading, nowMs: nowMs)
    }

    private func join(_ reading: PadReading, nowMs: Double) -> Int? {
        let seat = Self.seatId(forSlot: reading.slot)
        let name = coordinator.padName(reading.id)
        // The same HELLO a phone sends. autoName stays false: the pad's name is a
        // real (if borrowed) identity, not a request for an HX-n slot.
        coordinator.deliverLocal(from: seat, data: [
            "type": MSG.hello, "name": name, "autoName": false,
        ])
        // A refused join (room full) leaves no row behind. Drop the seat, and
        // note the refusal so the retry runs on the clock (see refusedAt) rather
        // than on every poll.
        guard coordinator.player(seat) != nil else {
            refusedAt[reading.slot] = nowMs
            return nil
        }
        refusedAt.removeValue(forKey: reading.slot)
        seated[reading.slot] = seat
        return seat
    }

    private func retireVanished(_ readings: [PadReading]) {
        let live = Set(readings.map(\.slot))
        // A replugged pad should try again at once, not wait out an old refusal.
        refusedAt = refusedAt.filter { live.contains($0.key) }
        for (slot, seat) in seated where !live.contains(slot) {
            seated.removeValue(forKey: slot)
            // Same path as a phone closing its tab: mid-game the row is held
            // (with a rejoin QR) so replugging the pad, or scanning with a phone,
            // resumes the seat; in lobby or results it is dropped outright.
            coordinator.deliverLocal(from: seat, data: ["type": MSG.leave])
        }
    }

    // MARK: - Menus

    /// The lobby steps this seat's start level on LEFT/RIGHT only. Up/down belong
    /// to the focus engine: the lobby's two controls are stacked vertically (START
    /// bottom-center, ⓘ top-right — see LobbyView), and nothing sits beside either,
    /// so the two axes divide cleanly with no input taken away from the UI.
    ///
    /// That split is what lets several pads coexist here. Level and colour are
    /// PER-PLAYER and never touch focus, so each pad sets its own; START and ⓘ are
    /// decisions about the display, where one shared ring is the same model the
    /// pause and results overlays already use.
    ///
    /// Rests on the lobby having nothing focusable side-by-side. Put two controls
    /// in a row there and left/right starts moving the ring as well as the level.
    private func onMenuNav(seat: Int, direction: String) {
        guard coordinator.state == .lobby else { return }
        let step: Int
        if direction == "right" { step = 1 }
        else if direction == "left" { step = -1 }
        else { return }   // up/down belong to the focus engine
        guard let level = coordinator.roomLevelAfterStep(seat: seat, delta: step) else { return }
        coordinator.deliverLocal(from: seat, data: ["type": MSG.setLevel, "level": level])
    }

    /// Only what the focus engine cannot reach on its own. Outside a running match
    /// the pad drives focus exactly like a remote, so START needs no binding here:
    /// the bottom face button is tvOS's Select and already clicks it. Binding it
    /// again would also start the round on a press aimed at ⓘ. Same reason Play
    /// Again and Continue are absent, and why a button added to any of these
    /// screens is pad-reachable the day it lands.
    private func onMenuPress(seat: Int, index: Int) {
        guard coordinator.state == .lobby else { return }

        // Index 9 is the exception, as everywhere (see route): a gamepad's Menu
        // button never reaches the responder chain, so if it is to mean anything
        // the binding must be here. In the lobby it starts the round — the same
        // host-gated Start the web and Android give it. Not the bottom face
        // button too (Android's other binding): here that is Select, and the
        // focus engine already clicks the focused START with it.
        if index == PadButton.start {
            if seat == coordinator.hostPeerIndex {
                coordinator.deliverLocal(from: seat, data: ["type": MSG.startGame])
            }
            return
        }

        // Colour has no on-screen control to focus (the picker lives on the
        // phone), so it keeps a shoulder side of its own in each direction. The
        // room core resolves which slot is next; this then sends the same
        // SET_COLOR the phone's picker sends.
        var step = 0
        if index == PadButton.l1 || index == PadButton.l2 { step = -1 }
        else if index == PadButton.r1 || index == PadButton.r2 { step = 1 }
        guard step != 0, let next = coordinator.roomColorAfterStep(seat: seat, step: step) else { return }
        coordinator.deliverLocal(from: seat, data: ["type": MSG.setColor, "colorIndex": next])
    }

    // MARK: - Rumble

    /// Effects a phone gets as haptics. Driven off engine events rather than a
    /// message, because the pad is local and the event is right there. WHAT each
    /// one feels like is the shared table's call, not this file's.
    public func handle(event: GameEvent) {
        guard !seated.isEmpty else { return }
        switch event.type {
        case "garbage_sent":
            guard let toId = event.toId else { return }
            rumble(seat: toId, "garbageSent", lines: event.lines ?? 0)
        case "garbage_cancelled":
            guard let playerId = event.playerId else { return }
            rumble(seat: playerId, "garbageCancelled")
        case "garbage_applied":
            // The stack just got shoved up. Distinct from garbage_sent, which is
            // only the telegraph: in between, this player could still have
            // cancelled it.
            guard let playerId = event.playerId else { return }
            rumble(seat: playerId, "garbageApplied", lines: event.lines ?? 0)
        case "line_clear":
            guard let playerId = event.playerId else { return }
            rumble(seat: playerId, "lineClear", lines: event.lines ?? 0)
        case "player_ko":
            guard let playerId = event.playerId else { return }
            rumble(seat: playerId, "playerKO")
        default:
            break
        }
    }

    /// Cached across the whole session, misses included: the effect for a given
    /// (kind, lines) never changes, and looking it up crosses the bridge.
    private var effects: [String: EngineBridge.PadRumble?] = [:]

    private func rumble(seat: Int, _ kind: String, lines: Int = 0) {
        guard let slot = seated.first(where: { $0.value == seat })?.key else { return }
        let key = "\(kind):\(lines)"
        let effect: EngineBridge.PadRumble?
        if let cached = effects[key] {
            effect = cached
        } else {
            effect = coordinator.padRumble(kind, lines: lines)
            // updateValue, not the subscript: subscript-assigning nil REMOVES
            // the key, so a miss would cross the bridge on every event (the
            // Android twin caches the miss too).
            effects.updateValue(effect, forKey: key)
        }
        guard let effect else { return }
        source.rumble(slot: slot, durationMs: effect.durationMs,
                      weak: effect.weak, strong: effect.strong)
    }
}

File: android/core/src/commonMain/kotlin/com/hexstacker/core/display/PadSeats.kt
package com.hexstacker.core.display

import com.hexstacker.core.engine.EngineBridge
import com.hexstacker.core.model.GameEvent
import com.hexstacker.core.net.Msg
import com.hexstacker.core.net.RoomState
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.double
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Gamepads attached to the TV itself, as players.
 *
 * A pad is not a second kind of player. Every press becomes the SAME message a
 * phone would have sent and goes through the coordinator's ordinary inbound path,
 * so joining, auto-naming, colour slots, host election, liveness and pause all
 * keep running their one implementation. What a local seat skips is the relay,
 * which is why its peer index comes from a range the relay will never hand out —
 * see [PadSeats.LOCAL_SEAT_BASE] for the range and the packed-frame constraint
 * that makes it a HIGH POSITIVE one.
 *
 * The MAPPING is not here. Which button rotates, how fast a held direction
 * repeats and how the stick scales a soft drop live in `server/PadMapper.js` and
 * cross through [EngineBridge.padPollJson], because the web display, tvOS and
 * this must agree about what a press means. What IS here is the part that needs a
 * shell: which pads exist, which seat each holds, and where a menu press goes.
 *
 * Reading the hardware is a [PadSource], injected, so the lifecycle and routing
 * below are testable with no controller attached. The Android implementation
 * lives in :tv, where InputDevice does.
 */
interface PadSource {
    /**
     * Attached pads in STABLE slot order: a pad must keep its slot across polls
     * and reclaim it after a reconnect, or it would be a new player every time.
     */
    fun pads(): List<PadReading>

    /** Effects a phone gets as haptics through its own vibrate() call. */
    fun rumble(slot: Int, durationMs: Long, amplitude: Double)
}

data class PadReading(
    val slot: Int,
    /** The product string, cleaned up through the shared rules into a name. */
    val id: String,
    /** W3C "standard" mapping order, so one mapper serves every platform. */
    val buttons: List<Boolean>,
    /** `[leftX, leftY]`, y POSITIVE DOWN as the web reports it. The mapper reads
     *  nothing past the left stick, so nothing else is carried. */
    val axes: List<Double>,
)

/**
 * Button indices routed by hand here. The rest are the mapper's business and are
 * never named. Kept in sync with `PAD_BTN` in server/PadMapper.js.
 */
private object PadButton {
    const val FACE_DOWN = 0
    const val L1 = 4
    const val R1 = 5
    const val L2 = 6
    const val R2 = 7
    const val START = 9
}

internal class PadSeats(
    private val source: PadSource,
    private val coordinator: DisplayCoordinator,
) {
    companion object {
        /**
         * An id the relay will never hand out (it owns slot 0 and gives out
         * 1..MAX). Mirrors LOCAL_SEAT_BASE in server/PadMapper.js, which owns the
         * definition and the reason it must be POSITIVE: a player id reaches us
         * inside PartyCore's packed frame, where every integer is one UTF-16 code
         * unit, and a negative one is unencodable — packFrame throws, so the first
         * frame of a match with a pad seated kills the game. Pinned by
         * tests/protocol-swift-parity.test.js.
         */
        const val LOCAL_SEAT_BASE = 900

        /**
         * Derived from the pad's own slot, so a reconnecting pad lands back on the
         * same seat: a reconnect, not a new player.
         */
        fun seatId(slot: Int): Int = LOCAL_SEAT_BASE + slot

        /**
         * True for a seat this display owns rather than one the relay handed out.
         * Every per-peer send has to ask, because there is nothing to send to.
         */
        fun isLocalSeat(peerIndex: Int): Boolean = peerIndex >= LOCAL_SEAT_BASE
    }

    /** slot -> seat id, for pads that have actually joined. Game thread only. */
    private val seated = mutableMapOf<Int, Int>()

    /**
     * The seated slots, as an immutable snapshot for readers on OTHER threads:
     * [holdsSeat] is asked from the UI thread's dispatchKeyEvent and [hasSeats]
     * from the activity's tick gate, while [seated] mutates on the game thread.
     * Rewritten (never mutated) after every seat change, so a cross-thread read
     * sees a coherent set — at worst one poll stale, which both callers tolerate.
     */
    @kotlin.concurrent.Volatile
    private var seatedSlots: Set<Int> = emptySet()

    /** Slots whose join was refused (room full). A held button would otherwise
     *  re-send HELLO through the bridge on every poll; cleared once every button
     *  on that pad is released, so the retry is a fresh press, as on the web. */
    private val refused = mutableSetOf<Int>()

    val hasSeats: Boolean get() = seatedSlots.isNotEmpty()

    /**
     * Whether the pad in [slot] has actually joined. A press from one that has not
     * is a JOIN and must not also act, which [poll] already enforces internally by
     * handing that press to the mapper as a baseline. :tv asks the same question
     * at the other door, for the press Compose would otherwise turn into a click
     * on whatever is focused — Play Again on the results screen being the one that
     * bites.
     */
    fun holdsSeat(slot: Int?): Boolean = slot != null && seatedSlots.contains(slot)

    /**
     * Seats that joined on the LAST collect. Their press still goes to the
     * mapper, which is the point: it becomes the baseline, so the button that
     * joined reads as already-down rather than as a fresh press. What must not
     * happen is acting on it, or the bottom face button would join and start the
     * round in one press. Spans collect to route, which is why it is not a local.
     */
    private val joinedNow = mutableSetOf<Int>()

    /**
     * One poll's pad states: retire vanished pads, seat new ones, stamp
     * liveness. The mapping itself is the shim's; the CALLER decides which
     * crossing carries it — [poll]'s own, or the playing tick's frame
     * ([EngineBridge.framePads]), which is what keeps a playing tick at one
     * evaluate.
     */
    suspend fun collectStates(): JsonArray {
        val readings = source.pads()
        retireVanished(readings)
        joinedNow.clear()
        return buildJsonArray {
            for (reading in readings) {
                val existing = seated.containsKey(reading.slot)
                val seat = seatFor(reading) ?: continue
                if (!existing) joinedNow.add(seat)
                // A local seat sends nothing over the wire, so nothing else proves
                // it is still there. Its presence in this poll IS the proof; without
                // it the liveness sweep would expire an idle player mid-game.
                coordinator.markLocalSeatSeen(seat)
                add(
                    buildJsonObject {
                        put("seat", JsonPrimitive(seat))
                        put("buttons", JsonArray(reading.buttons.map { JsonPrimitive(it) }))
                        put("axes", JsonArray(reading.axes.map { JsonPrimitive(it) }))
                    }
                )
            }
        }
    }

    /**
     * Route one poll's results — the menu edges and the rumble flag; the game
     * input never comes back (the shim feeds it to the engine itself).
     */
    suspend fun route(results: JsonArray, playing: Boolean) {
        for (result in results) {
            val seat = result.jsonObject["seat"]?.jsonPrimitive?.int ?: continue
            // The joining press is a baseline, never an action. See joinedNow.
            if (joinedNow.contains(seat)) continue

            // The one effect driven by what the PLAYER did rather than what
            // happened to them. Fired off the mapping rather than the resulting
            // lock event, so the thump lands with the press.
            if (result.jsonObject["hardDrop"]?.jsonPrimitive?.boolean == true) {
                rumble(seat, "hardDrop")
            }
            val pressed = result.jsonObject["pressed"]?.jsonArray.orEmpty().map { it.jsonPrimitive.int }
            if (playing) {
                if (pressed.contains(PadButton.START)) {
                    coordinator.deliverLocal(seat, msg(Msg.PAUSE_GAME))
                }
                continue
            }
            // The UNPAUSED countdown, and only it. :tv consumes pad input during
            // the 3-2-1 (it is the GAME screen, unpaused), so nothing else hears
            // index 9 there and the remote could pause the countdown while a pad
            // could not. Once PAUSED the pad is NOT consumed — countdown included —
            // so the press reaches MainActivity's KEYCODE_BUTTON_START instead;
            // binding it here as well would toggle twice on one press and put the
            // overlay straight back up.
            if (coordinator.state == RoomState.COUNTDOWN) {
                if (!coordinator.paused && pressed.contains(PadButton.START)) {
                    // The DIRECT toggle, never remoteTogglePause: this runs inside
                    // the action consumer, and the acked public path would enqueue
                    // an action the consumer can never reach — a self-deadlock
                    // that froze the app on a pad's Start press mid-countdown.
                    coordinator.togglePause()
                }
                continue
            }
            for (step in result.jsonObject["nav"]?.jsonArray.orEmpty()) {
                onMenuNav(seat, step.jsonPrimitive.content)
            }
            for (index in pressed) onMenuPress(seat, index)
        }
    }

    /**
     * Collect, map and route in one go — the path for every tick that is NOT
     * running a frame (lobby, results, countdown, paused). The playing tick
     * rides its mapping on the frame crossing instead (see [collectStates]).
     * [nowMs] is the monotonic clock the mapper measures DAS and the soft-drop
     * keepalive against, NOT wall time.
     */
    suspend fun poll(nowMs: Double, playing: Boolean) {
        val states = collectStates()
        if (states.isEmpty()) return

        val results = try {
            coordinator.padPoll(states.toString(), nowMs, playing)
        } catch (e: Throwable) {
            coordinator.reportPadError("padPoll", e)
            return
        }
        route(results, playing)
    }

    // --- Seat lifecycle ------------------------------------------------------

    private suspend fun seatFor(reading: PadReading): Int? {
        val existing = seated[reading.slot]
        if (existing != null) {
            // The row can disappear without the pad going anywhere: a session
            // reset clears the whole roster. Give the seat up so the next press
            // joins the new room instead of feeding a player that is gone.
            if (!coordinator.hasPlayer(existing)) {
                seated.remove(reading.slot)
                seatedSlots = seated.keys.toSet()
                return null
            }
            return existing
        }
        // Any press joins. Naming one button would leave a player who pressed a
        // different one with no feedback, and no letter is right on every brand.
        // There is no welcome screen to exclude: a TV goes straight to the lobby.
        if (!reading.buttons.contains(true)) {
            refused.remove(reading.slot)
            return null
        }
        if (reading.slot in refused) return null
        return join(reading)
    }

    private suspend fun join(reading: PadReading): Int? {
        val seat = seatId(reading.slot)
        val name = coordinator.padName(reading.id)
        // The same HELLO a phone sends. autoName stays false: the pad's name is a
        // real (if borrowed) identity, not a request for an HX-n slot.
        coordinator.deliverLocal(
            seat,
            buildJsonObject {
                put("type", JsonPrimitive(Msg.HELLO))
                put("name", JsonPrimitive(name))
                put("autoName", JsonPrimitive(false))
            }
        )
        // A refused join (room full) leaves no row behind. Drop the seat, and
        // remember the refusal until the button is released (see [refused]).
        if (!coordinator.hasPlayer(seat)) {
            refused.add(reading.slot)
            return null
        }
        seated[reading.slot] = seat
        seatedSlots = seated.keys.toSet()
        return seat
    }

    private suspend fun retireVanished(readings: List<PadReading>) {
        val live = readings.map { it.slot }.toSet()
        refused.retainAll(live)
        for ((slot, seat) in seated.entries.toList()) {
            if (live.contains(slot)) continue
            seated.remove(slot)
            seatedSlots = seated.keys.toSet()
            // Same path as a phone closing its tab: mid-game the row is held (with
            // a rejoin QR) so a returning pad, or a phone scanning, resumes the
            // seat; in lobby or results it is dropped outright.
            coordinator.deliverLocal(seat, msg(Msg.LEAVE))
        }
    }

    // --- Menus ---------------------------------------------------------------

    /**
     * The lobby steps this seat's start level, which is why :tv stops routing pad
     * D-pad presses into Compose focus there: the D-pad cannot both move a focus
     * ring and set a level. Everywhere else outside play the pad is left alone to
     * drive focus like a remote, so Play Again and Continue need no binding here.
     */
    private suspend fun onMenuNav(seat: Int, direction: String) {
        if (coordinator.state != RoomState.LOBBY) return
        val step = if (direction == "right" || direction == "up") 1 else -1
        val level = coordinator.roomCore.levelAfterStep(seat, step) ?: return
        coordinator.deliverLocal(
            seat,
            buildJsonObject {
                put("type", JsonPrimitive(Msg.SET_LEVEL))
                put("level", JsonPrimitive(level))
            }
        )
    }

    /**
     * The lobby's actions, bound here because the pad owns that screen's input and
     * Compose focus never sees these presses. On the overlays the opposite holds,
     * so Play Again and Continue are deliberately absent.
     */
    private suspend fun onMenuPress(seat: Int, index: Int) {
        if (coordinator.state != RoomState.LOBBY) return

        // Starting the round is the host's call, the same rule the phones' lobby
        // renders. The bottom face button because it is the one a player reaches
        // for when they want something to happen, and Start because its meaning
        // holds on every brand. See server/PadMapper.js for why no face button is
        // brand-safe and why that does not matter in a lobby.
        if (index == PadButton.FACE_DOWN || index == PadButton.START) {
            if (seat == coordinator.room.hostPeerIndex) coordinator.deliverLocal(seat, msg(Msg.START_GAME))
            return
        }

        // Colour has no on-screen control to focus (the picker lives on the
        // phone), so it keeps a shoulder side of its own in each direction. The
        // room core resolves which slot is next; this then sends the same
        // SET_COLOR the phone's picker sends.
        val step = when (index) {
            PadButton.L1, PadButton.L2 -> -1
            PadButton.R1, PadButton.R2 -> 1
            else -> return
        }
        val next = coordinator.roomCore.colorAfterStep(seat, step) ?: return
        coordinator.deliverLocal(
            seat,
            buildJsonObject {
                put("type", JsonPrimitive(Msg.SET_COLOR))
                put("colorIndex", JsonPrimitive(next))
            }
        )
    }

    // --- Rumble --------------------------------------------------------------

    /**
     * Effects a phone gets as haptics, driven off engine events rather than a
     * message because the pad is local and the event is right there. Android has
     * one amplitude per effect rather than two motors, so the dual-rumble weak and
     * strong magnitudes collapse to the stronger of the pair.
     */
    suspend fun handle(event: GameEvent) {
        if (seated.isEmpty()) return
        when (event.type) {
            "garbage_sent" -> event.toId?.let { rumble(it, "garbageSent", event.lines ?: 0) }
            "garbage_cancelled" -> event.playerId?.let { rumble(it, "garbageCancelled") }
            // The stack just got shoved up. Distinct from garbage_sent, which is
            // only the telegraph: in between, this player could still have
            // cancelled it.
            "garbage_applied" -> event.playerId?.let { rumble(it, "garbageApplied", event.lines ?: 0) }
            "line_clear" -> event.playerId?.let { rumble(it, "lineClear", event.lines ?: 0) }
            "player_ko" -> event.playerId?.let { rumble(it, "playerKO") }
            else -> Unit
        }
    }

    /**
     * Cached across the whole session: the effect for a given (kind, lines) never
     * changes, and looking it up crosses the bridge.
     */
    private val effects = mutableMapOf<String, Effect?>()

    private data class Effect(val durationMs: Long, val amplitude: Double)

    private suspend fun rumble(seat: Int, kind: String, lines: Int = 0) {
        val slot = seated.entries.firstOrNull { it.value == seat }?.key ?: return
        val key = "$kind:$lines"
        val effect = if (effects.containsKey(key)) effects[key] else {
            val decoded = runCatching {
                val json = coordinator.padRumbleJson(kind, lines)
                val obj = Json.Default.parseToJsonElement(json).jsonObject
                Effect(
                    durationMs = obj.getValue("durationMs").jsonPrimitive.double.toLong(),
                    // Android's InputDevice vibrator has ONE motor where a
                    // dual-rumble pad has two, so the pair collapses to the
                    // stronger of them rather than averaging into mush.
                    amplitude = maxOf(
                        obj.getValue("weak").jsonPrimitive.double,
                        obj.getValue("strong").jsonPrimitive.double,
                    ),
                )
            }.getOrNull()
            effects[key] = decoded
            decoded
        }
        effect?.let { source.rumble(slot, it.durationMs, it.amplitude) }
    }

    private fun msg(type: String): JsonObject =
        buildJsonObject { put("type", JsonPrimitive(type)) }
}
